use case:credentials: record Linux capability usage (#1189)
cilium/tetragon: docs/content/en/docs/use-cases/security-profiles/record-linux-capabilities.md

---
title: "Record Linux Capabilities Usage"
weight: 2
icon: "overview"
description: "Record a capability profile of pods and containers"
---

When the kernel needs to perform a privileged operation on behalf of a process, it checks
the [Capabilities](https://man7.org/linux/man-pages/man7/capabilities.7.html) of the process
and issues a verdict to allow or deny the operation.

Tetragon is able to record these checks performed by the kernel. This can be used to answer
the following questions:

> What is the capabilities profile of pods or containters running in the cluster?

> What capabilities to add or remove when [configuring a security context for a pod or container](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-capabilities-for-a-container)?

## Kubernetes Environments

First, verify that your k8s environment is set up and that all pods are up and running, and deploy the demo application:

```shell-session
kubectl create -f https://raw.githubusercontent.com/cilium/cilium/v1.11/examples/minikube/http-sw-app.yaml
```

It might take several seconds until all pods are Running:

> ```shell-session
> kubectl get pods -A
> NAMESPACE            NAME                                         READY   STATUS    RESTARTS   AGE
> default              deathstar-54bb8475cc-6c6lc                   1/1     Running   0          2m54s
> default              deathstar-54bb8475cc-zmfkr                   1/1     Running   0          2m54s
> default              tiefighter                                   1/1     Running   0          2m54s
> default              xwing                                        1/1     Running   0          2m54s
> kube-system          tetragon-sdwv6                               2/2     Running   0          27m
> ```

Reviewer: ````suggestion
```shell-session
kubectl get pods -A
```

The output should be similar to:
```
NAMESPACE            NAME                                         READY   STATUS    RESTARTS   AGE
default              deathstar-54bb8475cc-6c6lc                   1/1     Running   0          2m54s
default              deathstar-54bb8475cc-zmfkr                   1/1     Running   0          2m54s
default              tiefighter                                   1/1     Running   0          2m54s
default              xwing                                        1/1     Running   0          2m54s
kube-system          tetragon-sdwv6                               2/2     Running   0          27m
```
````

To reduce events, the [creds-capability-usage](https://raw.githubusercontent.com/cilium/tetragon/main/examples/tracingpolicy/process-credentials/creds-capability-usage.yaml) rate limits events to 1 minute. More details about rate-limiting can be found in the [tracing policy documentation]({{< ref "docs/concepts/tracing-policy#actions-filter" >}}).
{{< /note >}}

Apply the [creds-capability-usage](https://raw.githubusercontent.com/cilium/tetragon/main/examples/tracingpolicy/process-credentials/creds-capability-usage.yaml) policy:
```shell-session
kubectl apply -f https://raw.githubusercontent.com/cilium/tetragon/main/examples/tracingpolicy/process-credentials/creds-capability-usage.yaml
```

Start monitoring for events with `tetra` cli, but match only events of `xwing` pod:
```shell-session
kubectl exec -it -n kube-system ds/tetragon -c tetragon -- tetra getevents --namespaces default --pods xwing
```

In another terminal, kubectl exec into the xwing pod:
```shell-session
kubectl exec -it xwing -- /bin/bash
```

As an example execute [dmesg](https://man7.org/linux/man-pages/man1/dmesg.1.html) to print the kernel ring buffer. This requires the special capability `CAP_SYSLOG`:

> ```shell-session
> dmesg
> dmesg: klogctl: Operation not permitted
> ```

Reviewer: ````suggestion
```shell-session
dmesg
```

The output should be similar to:
```
dmesg: klogctl: Operation not permitted
```
````

    "args": [
      {
        "user_ns_arg": {
          "level": 0,
          "uid": 0,
          "gid": 0,
          "ns": {
            "inum": 4026531837,
            "is_host": true
          }
        }
      },
      {
        "capability_arg": {
          "value": 34,
          "name": "CAP_SYSLOG"
        }
      }
    ],
    "return": {
      "int_arg": -1
    },
    "action": "KPROBE_ACTION_POST"
  },
  "node_name": "kind-control-plane",
  "time": "2023-07-06T10:13:33.834882128Z"
}
```

In addition to the Kubernetes Identity and process metadata from exec events, [ProcessKprobe]({{< ref "/docs/reference/grpc-api#processkprobe" >}}) events contain the arguments of the observed system call. In the above case they are:

* `function_name`: that is the `cap_capable` kernel function.

* `user_ns_arg`: is the [user namespace]({{< ref "/docs/reference/grpc-api#usernamespace" >}}) where the capability is required.

   * `level`: is the nested level of the user namespace. Here it is zero which indicates the initial user namespace.
   * `uid`: is the user ID of the owner of the user namespace.
   * `gid`: is the group ID of the owner of the user namespace.

>    * [ns]({{< ref "/docs/reference/grpc-api#namespace" >}}): details the information about the namespace. `is_host` indicates that the target user namespace where the capability is required is the host namespace.

Reviewer: ```suggestion
   * [`ns`]({{< ref "/docs/reference/grpc-api#namespace" >}}): details the information about the namespace. `is_host` indicates that the target user namespace where the capability is required is the host namespace.
```

* `capability_arg`: is the [capability]({{< ref "/docs/reference/grpc-api#kprobecapability" >}}) required to perform the operation. In this example reading the kernel ring buffer.

   * `value`: is the integer number of the required capability.
   * `name`: is the name of the required capability. Here it is the `CAP_SYSLOG`.

> * `return`: indicates via the `int_arg` if the capability check succeeded or failed. `0` means it succeeded and the access was granted while `-1` means it failed and the operation was denied.

Reviewer: Usually, I would advise to not put space between list items as it can break markdown listing quite unexpectedly but it seems to work here.

To disable the [creds-capability-usage](https://raw.githubusercontent.com/cilium/tetragon/main/examples/tracingpolicy/process-credentials/creds-capability-usage.yaml) run:

> ```bash

Reviewer: ````suggestion
```shell-session
````

```shell-session
kubectl create -f https://raw.githubusercontent.com/cilium/cilium/v1.11/examples/minikube/http-sw-app.yaml
```

It might take several seconds until all pods are Running:

```shell-session
kubectl get pods -A
NAMESPACE            NAME                                         READY   STATUS    RESTARTS   AGE
default              deathstar-54bb8475cc-6c6lc                   1/1     Running   0          2m54s
default              deathstar-54bb8475cc-zmfkr                   1/1     Running   0          2m54s
default              tiefighter                                   1/1     Running   0          2m54s
default              xwing                                        1/1     Running   0          2m54s
kube-system          tetragon-sdwv6                               2/2     Running   0          27m
```

### Monitor Capability Checks

We use the [creds-capability-usage](https://raw.githubusercontent.com/cilium/tetragon/main/examples/tracingpolicy/process-credentials/creds-capability-usage.yaml) Tracing Policy that will generate a [ProcessKprobe]({{< ref "/docs/reference/grpc-api#processkprobe" >}}) event.

{{< note >}}

> Such Tracing Policies may emit a high number of events. Capability checks are performed on behalf of applications by the kernel when privileged operations are performed.
> 
> In order to reduce the number of events, the [creds-capability-usage](https://raw.githubusercontent.com/cilium/tetragon/main/examples/tracingpolicy/process-credentials/creds-capability-usage.yaml) rate limit the same event to 1 minute.

Reviewer 2: ```suggestion
Tracing policies as the one used here, may emit a high number of events.
To reduce events, the [creds-capability-usage](https://raw.githubusercontent.com/cilium/tetragon/main/examples/tracingpolicy/process-credentials/creds-capability-usage.yaml) rate limits events to 1 minute. More details about rate-limiting can be found in the [tracing policy documentation]({{< ref "docs/concepts/tracing-policy/" >}}).
```
